src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb

```diff
@@ -300,6 +298,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End Get
         End Property
 
+        Friend Overrides Function GetTypeDeclarationInfos(cancellationToken As CancellationToken) As ImmutableArray(Of TypeDeclarationInfo)
+            Return Declarations.GetTypeDeclarationInfos(Me, cancellationToken)
```

Comment:
instead of all this, i'm really curious if we could just expose the decl tree, just through an interface that is very minimal.  That would mean we could do this in a totally non-allocating fashion.

Reply:
I have been thinking about this approach too. Or maybe we can have a method on compilation like this
```cs
Task<ImmutableArray<T>>  VisitTopLevelTypeDelarations(Func<T, string, string, TypeKind, Accessibility, int>)
```